ibis/common/exceptions.py

```diff
@@ -127,3 +137,27 @@ def _mark_as_unsupported(self):
     _mark_as_unsupported.__name__ = f.__name__
 
     return _mark_as_unsupported
+
+
+class IbisUDFError(IbisError):
+    pass
+
+
+class MissingReturnAnnotationError(IbisUDFError):
+    def __str__(self):
+        (fn,) = self.args
+        signature = f"{fn.__qualname__}{inspect.signature(fn)}"
+        return f"function `{signature}` has no return type annotation"
+
+
+class MissingParameterAnnotationError(IbisUDFError):
+    def __str__(self):
+        fn, param = self.args
+        signature = f"{fn.__qualname__}{inspect.signature(fn)}"
+        return f"parameter `{param}` in function `{signature}` is missing a type annotation"
+
+
+class InvalidDecoratorError(IbisUDFError):
```

Comment:
Do we need to split these exceptions out into 3 different types. There's both a maintenance and user-understanding cost ("which exception should I be catching here?")  to having deep exception hierarchies.

If I was to do this I'd just use the base `IbisUDFError` everywhere (or maybe make a single subclass `IbisUDFDefinitionError` if there are non-signature-related errors that also need to be raised). I'd also perform the str formatting at exception instantiation time rather than in the `__str__` so only a str is attached to the instance. This is definitely part personal preference, but I prefer my exceptions to be simple data containers. Some exception tooling (for e.g. wrapping/chaining/serializing exceptions) also reasonably expects all exceptions to take a single str as their first argument - messing with this expectation can break things in hard to debug ways.

Reply:
This particular way of constructing exceptions is very deliberate.

It's based on my experience working with systems that are a mix of Rust and Python.

Based this experience, I've come to the conclusion that the typical way that we author (effectively no useful information beyond a string) and raise (raising incredibly generic exception types that again do nothing to help aid debugging `ValueError`, `TypeError`) exception types in Python results exceptions that are effectively impossible to debug without a debugger. It _is_ possible to better understand and debug an exception **without** the crutch of a debugger.

The way that is accomplished (at least to start) is to have more specific error types, that are capable of constructing error messages from a set of simple inputs (that aren't trivially a formatted string with the error).

[This blog post](https://kazlauskas.me/entries/errors) by a former colleague of mine captures the idea well, and shows how the same problem exists in Rust (and other languages besides Python!) and how to approach improvements. Notice that the specificity also helps with maintenance: you know what kind of thing went wrong at the call site by reading the exception type, rather than having to plumb through some format string.

This kind of approach is not without precedent in Python itself: the various OSError subclasses are an example of this approach.

> "which exception should I be catching here?"

I think the user-understanding part of this is a bit of a red herring. The exceptions we're talking about here are not something any end user of ibis should catch, they are problems to address (bugs in ibis, incorrect input from the user etc). Now, a library author might want to catch these, but IMO the situation is an improvement: you can now handle exceptions with more specificity. The tradeoff is, yeah, you have to know what those more specific exceptions are.

> Some exception tooling (for e.g. wrapping/chaining/serializing exceptions) also reasonably expects all exceptions to take a single str as their first argument.

I guess we'll just have to disagree here :) I don't think this is a reasonable assumption. There's nothing special about an exception that dictates this other than convention IIUC.

Now, all of this isn't without tradeoffs and there's almost certainly some level of specificity beyond which there's no additional utility.

I can give this another think through to see if there's a way to reduce the hierarchy while bringing in some good ideas from other communities.